include/fast_ber/ber_types/Real.hpp

```diff
@@ -1,61 +1,697 @@
+#include "fast_ber/util/SmallFixedIdBerContainer.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <regex>
 
 namespace fast_ber
 {
 
+namespace detail
+{
+constexpr size_t MaxEncodedLength = 100;
+
+/*enum class RealEncoding
+{
+    binary,
+    decimal,
+    special_real_value
+};
+
+enum class SpecialValues : uint8_t
+{
+    infinity       = 0b01000000,
+    minus_infinity = 0b01000001,
+    not_a_number   = 0b01000010,
+    minus_zero     = 0b01000011
+};*/
+
+} // namespace detail
+
+inline bool   decode_real(absl::Span<const uint8_t> input, double& output) noexcept;
+inline bool   decode_real_special(absl::Span<const uint8_t> input, double& output) noexcept;
+inline bool   decode_real_binary(absl::Span<const uint8_t> input, double& output) noexcept;
+inline bool   decode_real_decimal(absl::Span<const uint8_t> input, double& output) noexcept;
+inline bool   check_real_binary(absl::Span<const uint8_t> input) noexcept;
+inline size_t encode_real(absl::Span<uint8_t> output, double input) noexcept;
+
+static const std::regex iso_6093_nr1{"^[ ]*[+-]?[0-9]+$"};
+static const std::regex iso_6093_nr2{"^[ ]*[+-]?([0-9]+[\\.\\,]{1}[0-9]*|[0-9]*[\\.\\,]{1}[0-9]+)$"};
+static const std::regex iso_6093_nr3{"^[ ]*[+-]?([0-9]+[\\.\\,]{1}[0-9]*|[0-9]*[\\.\\,]{1}[0-9]+)[Ee][+-]?[0-9]+$"};
+
 template <typename Identifier = ExplicitId<UniversalTag::real>>
 class Real
 {
   public:
-    Real() noexcept = default;
+    Real() noexcept : m_contents{} {}
     Real(double num) noexcept { assign(num); }
-    Real(BerView& view) noexcept { assign_ber(view); }
-    Real(const Real&) noexcept;
-    Real(Real&&) noexcept;
-    ~Real() noexcept = default;
-
-    explicit Real(absl::Span<const uint8_t> ber_data) noexcept { assign_ber(ber_data); }
+    Real(BerView view) noexcept { decode(view); }
+    template <typename Identifier2>
+    Real(const Real<Identifier2>& rhs) noexcept;
 
     explicit operator double() const noexcept { return value(); }
     double   value() const noexcept;
 
-    Real&  operator=(double rhs) noexcept;
-    Real&  operator=(const Real& rhs) noexcept;
-    Real&  operator=(Real&& rhs) noexcept;
-    Real&  operator=(const BerView& rhs) noexcept;
-    void   assign(double val) noexcept;
-    void   assign(const Real& rhs) noexcept;
-    size_t assign_ber(const BerView& rhs) noexcept;
-    size_t assign_ber(absl::Span<const uint8_t> buffer) noexcept;
+    Real& operator=(double rhs) noexcept;
+    Real& operator=(const Real& rhs) noexcept;
+    Real& operator=(const BerView& rhs) noexcept;
+    template <typename Identifier2>
+    Real& operator=(const Real<Identifier2>& rhs) noexcept;
+
+    void assign(double val) noexcept;
+    void assign(const Real& rhs) noexcept;
 
-    bool operator==(const Real&) const
+    /* Inspired from https://www.embeddeduse.com/2019/08/26/qt-compare-two-floats/ */
+    bool operator==(const Real& rhs) const noexcept
```

Comment:
Move implementation for long function out of class declaration please

Reply:
Done